### ISSUE TITLE
Active Directory as an example for use of Social Media Login feature

### ISSUE THREAD

**Issue by lauracowen** [OpenLiberty/docs#2075]
Although the Social Media login was originally written for Google, GitHub etc type authentication, it's now actually just an easier way to do SSO authentication than using OIDC Client feature (which is why we tilt our concept docs towards Social Media Login rather than OIDC Client). However, although we talk about social media platforms, it's not clear that there are other SSO providers that can be used too (specifically Active Directory).

So, I think (see what you think or suggest other things if you think of them):

- Maybe add to the SSO intro topic "Active Directory" as an example provider (depending on wording, maybe need to say something like "as well as social media providers, can also use Active Directory as a provider" (I'm not certain of the wording so have a look and check with a security SME).
- Add the following config example to the Social Media Login feature to demonstrate (I think) how to configure other providers (check that with the security SME though):

  ```
    <oidcLogin
        id="liberty-aad-oidc-javaeecafe" clientId="${client.id}"
        clientSecret="${client.secret}"
        discoveryEndpoint="https://login.microsoftonline.com/${tenant.id}/v2.0/.well-known/openid-configuration"
        signatureAlgorithm="RS256" 
        userNameAttribute="preferred_username" />
  ```
(needs indentation tidying up)

**Comment by Rwalls1** [comment]
@ayoho Please review, I have added details regarding Active Directory as a provider: 
https://draft-openlibertyio.mybluemix.net/docs/20.0.0.11/reference/feature/socialLogin-1.0.html
https://draft-openlibertyio.mybluemix.net/docs/20.0.0.11/single-sign-on.html

**Comment by lauracowen** [comment]
The example config for AD in the Social Media Login feature looks fine, thanks.

In the SSO topic, I'm not sure it quite makes sense to have a separate Active Directory section because it's not equivalent to the other sections. You can use both SPNEGO and Social Media Login for AD connections. SPNEGO already mentions AD but the Social Media Login section (which is the focus of this issue) doesn't. The problem is (as you may have noticed) that the Social Media Login feature sounds like it's only for use with social media providers. However, it's actually just a more usable version of the OpenID Connect Client (another Liberty feature) so it can be used with AD as a provider as well; the Social Media Login feature is the one we want people to use (rather than OIDC Client). We need to make that clear in this SSO topic.

So, AD is a provider, like a social media platform. It's not (in the context of this topic) a method of connecting from Liberty to an AD server - you have to use SPNEGO or Social Media Login feature for that.

So, I don't think we need to explain what AD is (sorry - it's a nice description you've got). So instead of the separate AD section, I think it just needs another paragraph (couple of sentences) in the Social media login section explaining that you can configure connections to AD using the social media login SSO method (and link for example config to the example - directly if you can - in https://draft-openlibertyio.mybluemix.net/docs/20.0.0.11/reference/feature/socialLogin-1.0.html).

**Comment by lauracowen** [comment]
Can you use hardcoded names instead of variables for the bits in curly brackets? eg `${tenant.id}` - so as to give an example rather than a syntax format. The "Feature configuration elements" section beneath the Examples section provides a link to the `oidcLogin` element which lists all its attributes and what values can be put there.

Otherwise, looks good, thanks.

**Comment by Rwalls1** [comment]
@lauracowen To clarify, can you provide an example of a hardcoded name or value that would replace information in the curly brackets?

**Comment by lauracowen** [comment]
sorry, I don't know. You'll need someone from the security team to give an example tenant ID (not sure what the format would be).

**Comment by ayoho** [comment]
Looks like Microsoft's documentation (https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-protocols-oidc#fetch-the-openid-connect-metadata-document or https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-v2-protocols#endpoints) provides some examples of what that tenant ID string could be (e.g. `"common"`, `"organizations"`, `"consumers"`, or the "friendly domain name of the Azure AD tenant or the tenant's GUID identifier").

**Comment by Rwalls1** [comment]
@lauracowen Please review: https://draft-openlibertyio.mybluemix.net/docs/21.0.0.1/reference/feature/socialLogin-1.0.html

**Comment by lauracowen** [comment]
Could you add a sentence underneath the example saying something like "For more info about Active Directory endpoints, see the [Active Directory documentation](https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-v2-protocols#endpoints)"? So the user can see what "organizations" means in the string and can work out what value they should themselves use for that attribute.

Otherwise, looks fine, thank you.

**Comment by Rwalls1** [comment]
@lauracowen Thanks, please review the latest update: https://draft-openlibertyio.mybluemix.net/docs/21.0.0.1/reference/feature/socialLogin-1.0.html

**Comment by lauracowen** [comment]
Great, thank you! Signing off.

**Comment by dmuelle** [comment]
## Peer review
I found the reference to OIDC in the second sentence confusing, had to read it a couple times to realize that you configure AD login by using the `oidcLogin` element. I think you could add a sentence to explain. Possibly  something like:

"You can configure an Open Liberty server to use Active Directory as an authentication provider for protected resources. The `oidcLogin` element configures a social login by using the OpenId Connect protocol. With Open ID Connect, the discovery endpoint provides the information that a client needs to interact with the authentication provider, which in this case is Active Directory. In the following example, the `discoveryEndpoint` attribute specifies the endpoint for Active Directory:"

**Comment by Rwalls1** [comment]
@dmuelle Thanks, I've addressed your feedback.

**Comment by dmuelle** [comment]
Hi Richard,  one more minor suggestion- the last sentence reads a little strangely to me:

The following example shows the discoveryEndpoint attribute specify the endpoint for Active Directory:

It's not grammatically incorrect, but I think the dependent clause "the discoveryEndpoint attribute specify the endpoint for Active Directory" is hard to parse at a glance. I'd recommend:

"In the following example, the discoveryEndpoint attribute specifies the endpoint for Active Directory:"

otherwise looks good

**Comment by Rwalls1** [comment]
@dmuelle Thanks, I've addressed your feedback: https://draft-openlibertyio.mybluemix.net/docs/21.0.0.3/reference/feature/socialLogin-1.0.html

**Comment by dmuelle** [comment]
Sorry, one more thing I should've caught. 


"the discovery endpoint provides the..."
--->
"the discovery endpoint URL provides the..."
and
"the discoveryEndpoint attribute specifies the endpoint for"
--->
"the discoveryEndpoint attribute specifies the endpoint URL for"

**Comment by chirp1** [comment]
Hi Richard, Another nice example config! I ended up reviewing it for all the config examples instead of just Active Directory. Here are my comments.
- I believe the google ID for the "Request log in..." section is represented by "your_app_id" for clientId. Would you work in "your_app_id" and/or clientId into your description to give the user more context?
- "with the available providers" seems like it is missing some info/dangling. Maybe change "with the available providers" to "with the available providers included on the form.
- Globally change "app" to "application".
- Change "...so that users can have the option of logging in with ..." to "...so that users can log in with ..." to reduce wordiness.
- Rewrite to remove "out-of-the-box" from the "Provide other social media logins..." section, and which is flagged by Acrolinx. 
- Rerun Acrolinx and fix items as appropriate.
- For "For more information, see the authFilter configuration element.", fix the broken link. Also, we've found that highlghting in links is confusing. So, remove the highlighting from "authfFilter".
- The same goes for oauth2Login and oidcLogin 
- Change "The socialLogin-1.0 feature can be configured " to "The socialLogin feature can be configured " and remove the highlighting. A new version of the feature can become available. So, we don't want to specify a particular version, if possible.

**Comment by Rwalls1** [comment]
Because this feedback isn't related to the "Active Directory" section, I'm closing this issue and adding this feedback to a separate doc bug issue.
